### ISSUE TITLE
`TypedDict` errors on use of merge operator `|=` even if keys are correct

### ISSUE THREAD

**Issue by RenDelaCruz** [python/mypy#16244]
**Bug Report**

For `TypedDict`, mypy accepts the use of `.update()` using a `dict` with appropriate keys. However, using the merge operator `|=` for `TypedDict` gives an error.

**To Reproduce**

```python
class Foo(TypedDict):
    key: int
    
foo: Foo = {
    "key": 1
}

foo.update({"key": 2}) # Ok!

foo |= {"key": 3}
# error: Incompatible types in assignment (expression has type "dict[str, object]", variable has type "Foo")  [assignment]
```

[mypy Playground](https://mypy-play.net/?mypy=latest&python=3.11&gist=b1c0b89b3768464b6a15b3c66b88c387)

**Expected Behavior**

Using merge operator `|=` on `TypedDict` should be allowed, since `.update()` is supported.

**Actual Behavior**

Errors when using `|=` on `TypedDict`, even though the given keys are appropriate. This effectively means that projects using mypy cannot use [PEP 584](https://peps.python.org/pep-0584/) syntax.

**Comment by erictraut** [comment]
I think you meant for the last line in the code sample to be `foo |= {"key": 3}`. The value needs to be an `int`. With this fix, mypy still generates an error (a false positive).

**Comment by RenDelaCruz** [comment]
Yup sorry, edited now from `{"key": "3"}` to `{"key": 3}`.

But yes, mypy still generates an error even though `.update()` is accepted.
